contributors/design-proposals/alternate-api-representations.md

```diff
@@ -0,0 +1,405 @@
+
+To send ListOptions in the v2 future format, where the serialization of `resourceVersion`
+is changed to `rv`, clients would provide:
+
+    GET /api/v1/namespaces/example/replicasets?apiVersion=api.k8s.io/v2&rv=10
+
+Before we introduce a second API group version, we would have to ensure old servers
+properly reject apiVersions they do not understand.
+
+
+### Impact on web infrastructure
+
+In the past, web infrastructure and old browsers have coped poorly with the `Accept`
+header.  However, most modern caching infrastructure properly supports `Vary: Accept`
+and caching of responses has not been a significant requirement for Kubernetes APIs
+to this point.
+
+
+### Considerations for discoverability
+
+To ensure clients can discover these endpoints, the Swagger and OpenAPI documents
+should also include a set of example mime-types for each endpoint that are supported.
+Specifically, the `produces` field on an individual operation can be used to list a
+set of well known types. The description of the operation can include a stanza about
+retrieving alternate representations.
+
+
+## Alternatives considered
+
+*  Implement only with query parameters
+
+   To properly implement alternative resource versions must support multiple version
+   support (ask for v2, then v1). The Accept mechanism already handles this sort of
+   multi-version negotiation, while any approach based on query parameters would
+   have to implement this option as well. In addition, some serializations may not
+   be valid in all content types, so the client asking for TabularOutput in protobuf
+   may also ask for TabularOutput in JSON - if TabularOutput is not valid in protobuf
+   the server call fall back to JSON.
+
+*  Use new resource paths - `/apis/autoscaling/v1/namespaces/example/horizontalpodautoscalermetadata`
+
+   This leads to a proliferation of paths which will confuse automated tools and end
+   users. Authorization, logging, audit may all need a way to map the two resources
+   as equivalent, while clients would need a discovery mechanism that identifies a
+   "same underlying object" relationship that is different from subresources.
+
+*  Use a special HTTP header to denote the alternative representation
+
+   Given the need to support multiple versions, this would be reimplementing Accept
+   in a slightly different way, so we prefer to reuse Accept.
+
+*  For partial object retrieval, support complex field selectors
+
+   From an efficiency perspective, calculating subpaths and filtering out sub fields
+   from the underlying object is complex.  In practice, almost all filtering falls into
+   a few limited subsets, and thus retrieving an object into a few known schemas can be made
+   much more efficient. In addition, arbitrary transformation of the object provides
+   opportunities for supporting forward "partial" migration - for instance, returning a
+   ReplicationController as a ReplicaSet to simplify a transition across resource types.
```

Comment:
Also it's a bad idea to generalize without a few concrete implementations. We can always add this later, but adding it now would sign us up for a lot of maintenance costs.